HW2/P2/parallel_vector.pyx

```diff
@@ -107,9 +111,22 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):
                if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+                   # Add locks if they do not belong to the same lock, reduces locking for correlated data
+                   if dest[idx]/N==src[idx]/N:
+                     acquire(&locks[dest[idx]/N])
+                     counts[dest[idx]] += 1
+                     counts[src[idx]] -= 1
+                     release(&locks[src[idx]/N])
+
+                   else:
+                     acquire(&locks[dest[idx]/N])
+                     counts[dest[idx]] += 1
+                     release(&locks[dest[idx]/N])
+                     acquire(&locks[src[idx]/N])
+                     counts[src[idx]] -= 1
+                     release(&locks[src[idx]/N])
```

Comment:
Looks similar-ish, do you impose an order here (i.e. check if src[idx] > dest[idx]) in order to determine the order in which you lock things? I found that necessary for mine (looked up dining philosophers problem). The last pull request I reviewed did it more similarly to yours though so this may work just fine.